### PR TITLE
Stata lexer added

### PR DESCRIPTION
**Pull request by reifjulian**
I created a lexer for the [Stata](https://www.stata.com/) programming language. This PR includes demo, visual sample, and spec files.

### CODE REVIEW

File: spec/visual/samples/stata [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+	replace `v' = r(p50) if mi(`v')
+}
+compress
+
+foreach rhs in "mpg" "mpg weight" {
+	reg price `rhs' if foreign=="Domestic":origin, robust
+	reg price `rhs' if foreign=="Foreign":origin, robust
+}
+
+** EOF
```
Comment (tancnle) [inline spec/visual/samples/stata:168]: **suggestion:** Could we ensure there is a new line at the end of this file?

**Review by tancnle — CHANGES_REQUESTED** [review]
Thank you for your contribution @reifjulian ❤️ The PR looks good overall. I have left some suggestions for your consideration. Please let me know what you think 🙏🏼

File: spec/visual/samples/stata (hunk 1 above)
Comment (reifjulian) [reply spec/visual/samples/stata]: My January 10 commit has now addressed this issue.

File: lib/rouge/lexers/stata.rb [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+          elsif KEYWORD_TYPES.include? m[0]
+            token Keyword::Type
+          else
+            token Name
+          end
+		end
+
+        rule %r/[\[\]{}();,]/, Punctuation
+
+		rule %r([-<>?*+'^/\\!#.=~:&|]), Operator
```
[the remark this answers is not in the document]
Comment (reifjulian) [reply lib/rouge/lexers/stata.rb]: My January 10 commit has now addressed this issue.

File: lib/rouge/lexers/stata.rb [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+      desc "The Stata programming language (www.stata.com)"
+      tag 'stata'
+      filenames '*.do', '*.ado'
+	  mimetypes 'application/x-stata', 'text/x-stata'
+
+      # Note: types `str1-str2045` handled separately below
+      KEYWORD_TYPES = %w(byte int long float double str strL numeric string integer scalar matrix local global numlist varlist newlist)
+	  	  
+      # Stata commands used with braces
+	  KEYWORDS_RESERVED = %w(if else foreach forv forva forval forvalu forvalue forvalues to while in of continue break nobreak)
```
[the remark this answers is not in the document]
Comment (reifjulian) [reply lib/rouge/lexers/stata.rb]: Those keywords (e.g., `forv`, `forvalu`) are not typos; they are valid abbreviations for the command `forvalues`. My January 10 commit has added a comment on this point in this file. In addition, I have added a comment that includes a link to documentation for the Stata command.

File: lib/rouge/lexers/stata.rb [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+  module Lexers
+    class Stata < RegexLexer
+	  title "Stata"
+      desc "The Stata programming language (www.stata.com)"
+      tag 'stata'
+      filenames '*.do', '*.ado'
+	  mimetypes 'application/x-stata', 'text/x-stata'
+
+      # Note: types `str1-str2045` handled separately below
+      KEYWORD_TYPES = %w(byte int long float double str strL numeric string integer scalar matrix local global numlist varlist newlist)
```
[the remark this answers is not in the document]
Comment (reifjulian) [reply lib/rouge/lexers/stata.rb]: My January 10 commit has edited this code to use class methods.

File: lib/rouge/lexers/stata.rb (hunk 3 above)
[the remark this answers is not in the document]
Comment (reifjulian) [reply lib/rouge/lexers/stata.rb]: My January 10 commit has edited this code to use class methods.

File: lib/rouge/lexers/stata.rb [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+	  	  
+      # Stata commands used with braces
+	  KEYWORDS_RESERVED = %w(if else foreach forv forva forval forvalu forvalue forvalues to while in of continue break nobreak)
+
+      # Partial list of common programming and estimation commands, as of Stata 16
+	  # Note: not all abbreviations are included
+	  KEYWORDS = %w(
```
[the remark this answers is not in the document]
Comment (reifjulian) [reply lib/rouge/lexers/stata.rb]: These keywords are used in a regular expression:
[`rule %r/^\s*(#{KEYWORDS.join('|')})\b/, Keyword`](https://github.com/reifjulian/rouge/blob/7802afe0eee583355150be14a13fe7942ec17238/lib/rouge/lexers/stata.rb#L108)

My understanding from reading the [guidelines](https://github.com/rouge-ruby/rouge/blob/master/docs/LexerDevelopment.md#L287) is that one does not use class methods in this situation.

File: lib/rouge/lexers/stata.rb [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+		drawnorm edit encode decode erase expand export filefilter fillin format frame frget frlink gsort 
+		import dbase delimited excel fred haver sas sasxport5 sasxport8 spss infile infix input insobs inspect ipolate isid
+		joinby label language labelbook lookfor memory mem merge mkdir mvencode notes obs odbc order outfile
+		pctile xtile _pctile putmata range recast recode rename group reshape rm rmdir sample save saveold separate shell snapshot sort split splitsample stack statsby sysuse 
+		type unicode use varmanage vl webuse xpose zipfile 
+		number keep tab table tabulate stset stcox tsset xtset
+	  )
+	  
+      # Complete list of functions by name, as of Stata 16
+      PRIMITIVE_FUNCTIONS = %w(
```
[the remark this answers is not in the document]
Comment (reifjulian) [reply lib/rouge/lexers/stata.rb:57]: Similar to above, these are used in a regular expression:
[rule %r/\b(#{PRIMITIVE_FUNCTIONS.join('|')})(?=\()/, Name::Function](https://github.com/reifjulian/rouge/blob/stata/lib/rouge/lexers/stata.rb#L136)

**Comment by reifjulian** [comment]
Thank you for your helpful feedback, @tancnle. I have pushed a commit that incorporates most of your suggestions--see my specific responses above. Please let me know if you have further comments. This is my first time writing code in Ruby so your help is greatly appreciated :slightly_smiling_face:

**Comment by tancnle** [comment]
@reifjulian Thank you for considering these suggestions 🙇🏼 I have pushed a commit to your branch

- Rename the class methods, we generally use lowercase for method names
- Move constant declarations above class method definition, just styling reference
- Fix a small typo in the keywords
- Remove trailing spaces

Please let me know what you think. I am keen to merge it if you are happy.

**Comment by reifjulian** [comment]
@tancnle Your changes look good to me. I pushed a commit that made minor edits to comments. Please feel free to go ahead and merge. Thanks for reviewing this, I really appreciate it!

**Comment by tancnle** [comment]
Thank you @reifjulian 🚀

**Pull request merged**